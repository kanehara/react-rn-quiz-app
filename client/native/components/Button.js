import styled from 'styled-components'
import {View} from 'react-native'
import Text from 'native/components/Text'
import React from 'react'

const TouchableHighlight = styled.TouchableHighlight`
  background-color: ${({ theme }) => theme.buttonColor};
  border-radius: ${({theme}) => theme.buttonBorderRadius};
  border: ${({theme}) => theme.buttonBorder};
  
  width: 100%;
  align-items: center;
  justify-content: center;
  padding: 5px 10px;
  margin: 10px;
`

const ButtonText = styled(Text)`
  font-weight: 700;
`

const Button = props => (
  <TouchableHighlight 
    style={props.style} 
    onPress={props.onPress}
    underlayColor='rgba(0, 0, 0, 0.15)'
  >
    <View>
      <ButtonText>{props.title}</ButtonText>
    </View>
  </TouchableHighlight>
)

export const FalseButton = styled(Button)`
  background-color: ${({theme}) => theme.falseButtonColor};
`

export const TrueButton = styled(Button)`
  background-color: ${({theme}) => theme.trueButtonColor};
`

export default Button
