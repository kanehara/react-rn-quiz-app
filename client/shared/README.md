# The shared module for the trivia clients
A module housing common JS code between the web and React Native clients.

## Contains
* Redux actions, reducers, getters
* Apollo client used in redux layer to fetch data
* styled components theme