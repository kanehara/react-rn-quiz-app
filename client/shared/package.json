{
  "name": "shared",
  "version": "0.0.1",
  "scripts": {
    "test": "jest"
  },
  "devDependencies": {
    "@babel/core": "^7.2.2",
    "@babel/preset-env": "^7.3.1",
    "babel-jest": "^24.0.0",
    "jest": "^24.0.0",
    "unfetch": "^4.0.1"
  }
}
