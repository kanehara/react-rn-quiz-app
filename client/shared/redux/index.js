import { createStore, applyMiddleware, compose } from 'redux'
import thunk from 'redux-thunk'
import { quizReducer } from './quiz'
import { persistStore, persistReducer } from 'redux-persist'
import storage from 'redux-persist/lib/storage'

const composeEnhancers =
  typeof window === 'object' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ 
    ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({}) 
    : compose

const persistedReducer = persistReducer({
  key: 'root',
  storage
}, quizReducer)

const store = createStore(
  persistedReducer,
  composeEnhancers(
    applyMiddleware(thunk)
  )
)

const persistor = persistStore(store)

export {
  store,
  persistor
}
