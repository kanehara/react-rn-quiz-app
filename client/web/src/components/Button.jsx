import styled from 'styled-components'

const Button = styled.button`
  padding: 10px;
  width: 100%;
  font-size: 18px;
  border-radius: ${({theme}) => theme.buttonBorderRadius};
  border: ${({theme}) => theme.buttonBorder};
  font-weight: bold;
  background-color: ${props => props.theme.buttonColor};

  &:hover {
    cursor: pointer;
  }
`

export const TrueButton = styled(Button)`
  background-color: ${({theme}) => theme.trueButtonColor};
`

export const FalseButton = styled(Button)`
  background-color: ${({theme}) => theme.falseButtonColor};
`

export default Button
