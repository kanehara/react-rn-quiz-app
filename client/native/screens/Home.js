import React from 'react'
import {Button, ActivityIndicator} from 'react-native'
import Text from 'native/components/Text'
import { H1, H2 } from 'native/components/Header'
import ScreenView from 'native/screens/ScreenView'
import { connect } from 'react-redux'
import { actions, getters } from 'shared/redux/quiz'

class Home extends React.Component {
  state = {
    loadingQuiz: false,
    error: null
  }

  componentDidMount () {
    if (this.props.inProgress) {
      this.props.navigation.navigate('quiz')
    } else if (this.props.finished) {
      this.props.navigation.navigate('results')
    }
  }

  handleStartPress = async () => {
    try {
      this.setState({ loadingQuiz: true })
      await this.props.startQuiz()
      this.props.navigation.navigate('quiz')
    } catch (e) {
      this.setState({ error: e, loadingQuiz: false })
    }
  }

  render() {
    return (
      <ScreenView>
        {this.state.loadingQuiz 
          ? <ActivityIndicator />
          : (
            <React.Fragment>
              <H1>Welcome to the Trivia Challenge!</H1>
              <H2>You will be presented with 10 True or False questions.</H2>
              <Text>Can you score 100%?</Text>
              {this.state.loadingQuiz 
                ? <Text>Loading...</Text>
                : <Button onPress={this.handleStartPress} title='BEGIN' />
              }
              {this.state.error && <Text>Something went wrong while fetching questions.</Text>}
            </React.Fragment>
          )
        }
      </ScreenView>
    );
  }
}

export default connect(
  state => ({
    finished: getters.finished(state),
    inProgress: getters.inProgress(state)
  }),
  dispatch => ({
    startQuiz: () => dispatch(actions.startQuiz())
  })
)(Home)
