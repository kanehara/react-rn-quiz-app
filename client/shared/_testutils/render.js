import React from 'react'
import { ThemeProvider } from 'styled-components'
import theme from 'shared/theme'
import { Provider } from 'react-redux' 
import { store } from 'shared/redux'
import renderer from 'react-test-renderer'

export const render = jsx => renderer.create(
  <Provider store={store}>
    <ThemeProvider theme={theme}>
      {jsx}
    </ThemeProvider>
  </Provider>
)

