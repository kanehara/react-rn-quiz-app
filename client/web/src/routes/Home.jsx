import React from 'react'
import { Ray } from "rayout"
import Button from '../components/Button'
import { connect } from 'react-redux'
import { actions, getters } from 'shared/redux/quiz'

export class Home extends React.Component {
  state = {
    loading: false,
    error: null
  }

  componentDidMount () {
    if (this.props.inProgress) {
      this.props.history.replace('/quiz')
    } else if (this.props.finished) {
      this.props.history.replace('/results')
    }
  }

  startQuiz = async () => {
    try {
      this.setState({loading: true})
      await this.props.startQuiz()
      this.props.history.push('/quiz')
    } catch(e) {
      this.setState({ error: e, loading: false })
    }
  }

  render () {
    return (
      <Ray>
        <h1>Welcome to the Trivia Challenge!</h1>
        <h3>You will be presented with 10 True or False questions.</h3>
        <h3>Can you score 100%?</h3>
        {this.state.error 
          ? (
            <React.Fragment>
              <h4>Something went wrong while fetching your quiz questions.</h4>
              <Button onClick={this.startQuiz}>Try again?</Button>
            </React.Fragment>
          )
          : <Button onClick={this.startQuiz}>BEGIN</Button>
        }
      </Ray>
    )
  }
}

const mapStateToProps = state => ({
  finished: getters.finished(state),
  inProgress: getters.inProgress(state)
})

const mapDispatchToProps = (dispatch) => ({
  startQuiz: () => dispatch(actions.startQuiz())
})

export default connect(mapStateToProps, mapDispatchToProps)(Home)
