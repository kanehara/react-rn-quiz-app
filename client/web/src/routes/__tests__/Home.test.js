import {Home} from '../Home'
import React from 'react'
import { render } from 'shared/_testutils'
import Button from 'web/src/components/Button'

describe('Home', () => {
  const getMockHistory  = () => ({
    replace: jest.fn(),
    push: jest.fn()
  })

  it('renders', () => {
    expect(render(<Home />)).toMatchSnapshot()
  })
  
  it('replaces to quiz if in progress', () => {
    const mockHistory = getMockHistory()
    render(<Home history={mockHistory} inProgress />)
    expect(mockHistory.replace).toHaveBeenCalledWith('/quiz')
  })
  
  it('replaces to results if finished', () => {
    const mockHistory = getMockHistory()
    render(<Home history={mockHistory} finished />)
    expect(mockHistory.replace).toHaveBeenCalledWith('/results')
  })
  
  it('navigates to quiz on start', async () => {
    const mockHistory = getMockHistory()
    const startQuizMockPromise = Promise.resolve()
    const startQuizMock = jest.fn(() => startQuizMockPromise)
    const instance = render(<Home history={mockHistory} startQuiz={startQuizMock} />)
    const button = instance.root.findByType(Button)
    button.props.onClick()

    await startQuizMockPromise
    expect(startQuizMock).toHaveBeenCalled()
    expect(mockHistory.push).toHaveBeenCalledWith('/quiz')
  })
})
