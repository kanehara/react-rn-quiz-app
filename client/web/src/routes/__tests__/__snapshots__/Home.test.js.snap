// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders 1`] = `
<div
  className="sc-bdVaJa gRiozE"
>
  <h1>
    Welcome to the Trivia Challenge!
  </h1>
  <h3>
    You will be presented with 10 True or False questions.
  </h3>
  <h3>
    Can you score 100%?
  </h3>
  <button
    className="sc-bdVaJa laBHJB"
    onClick={[Function]}
  >
    BEGIN
  </button>
</div>
`;
